### DISCUSSION TITLE
What do to with `finalize` in our Java code

### DISCUSSION THREAD

**Discussion by iboB** [alpaca-core/alpaca-core-sdk#116] [Dev]
`finalize` has been deprecated for a long time and marked for removal recently. Some java implementations threaten to stop calling finalize altogether in the future. Our code relies on it and, in truth, since we wrap `shared_ptr`, it's actually a good fit for us in spite of all discussions about how using finalizers in Java is bad. 

What can we do:

* Ignore this and [suppress the warnings](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/SuppressWarnings.html). Our main target platform Android does implement finalizers. All jvm implementations still do implement it. As mentioned it *is* actually a good fit for our particular use case.
* Use [`Cleaner`](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html). It basically runs a thread in the background which calls callbacks (manually registered by us) when our objects become unreachable.
* Use [Phantom references](https://docs.oracle.com/javase/8/docs/api/java/lang/ref/PhantomReference.html). It's the most powerful way of dealing with manual cleanup in Java. It requires much more boilerplate than `Cleaner` but in this case we control the disposal thread ourselves.
* Something else?

**Comment by IvanFilipov** [comment]
With little to no expertise on the topic & a few minutes of reading the links, my vote goes to [Cleaner](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html).

**Comment by iboB** [comment]
This topic is now mute, as we decided to go for manual resource management
